File: CPPVersion/AoC2020CPP/AoC2020CPP/Day12.cpp
#include "Day12.h"
#include "UsefulMethods.h"

Day12::Day12() :DefaultClass(12, true)
{
	inp = FR.ReadLine();
	Part1();
	TimeResetClock();
	Part2();
	TimeClock();
}

int Day12::DegreeToTurn(int degree)
{
	if (degree == 90)
		return 1;
	else if (degree == 180)
		return 2;
	else if (degree == 270)
		return 3;
	return 0;
}

void Day12::Part1()
{
	int x = 0, y = 0;
	int facing = 1;
	int num;
	for (int i = 0; i < inp.size(); i++)
	{
		num = std::stoi(inp[i].substr(1));
		switch (inp[i][0])
		{
		case('F'):
			if (facing == 0)
				y += num;
			else if (facing == 1)
				x += num;
			else if (facing == 2)
				y -= num;
			else if (facing == 3)
				x -= num;
			break;
		case('R'):
			facing += DegreeToTurn(num);
			if (facing > 3)
				facing -= 4;
			break;
		case('L'):
			facing -= DegreeToTurn(num);
			if (facing < 0)
				facing += 4;
			break;
		case('N'):
			y += num;
			break;
		case('S'):
			y -= num;
			break;
		case('E'):
			x += num;
			break;
		case('W'):
			x -= num;
			break;
		}
	}
	std::cout << "Part1: " << UsefulMethods::ManhattenDistance(0, 0, x, y)<<"\n";
}

void Day12::Part2()
{
	int x = 0, y = 0;
	int Wx = 10, Wy = 1;
	int facing = 1;
	int num;
	for (int i = 0; i < inp.size(); i++)
	{
		num = std::stoi(inp[i].substr(1));
		switch (inp[i][0])
		{
		case('F'):
			for (int c = 0; c < num; c++)
			{
				x += Wx;
				y += Wy;
			}
			break;
		case('R'):
			for (int c = 0; c < DegreeToTurn(num); c++)
			{
				int temp = Wx;
				Wx = Wy;
				Wy = -1*temp;
			}
			break;
		case('L'):
			for (int c = 0; c < DegreeToTurn(num); c++)
			{
				int temp = Wx;
				Wx = -1 * Wy;
				Wy = temp;
			}
			break;
		case('N'):
			Wy += num;
			break;
		case('S'):
			Wy -= num;
			break;
		case('E'):
			Wx += num;
			break;
		case('W'):
			Wx -= num;
			break;
		}
	}
	std::cout << "Part1: " << UsefulMethods::ManhattenDistance(0, 0, x, y) << "\n";
}

File: JavaVersion/AoC2020/src/com/company/Day12.java
package com.company;

import java.util.ArrayList;

public class Day12 extends DefaultClass{

    public Day12()
    {
        super(12,true);
        int x=0,y=0;
        int facing = 1;
        ArrayList<String> inp =fr.ReadLine();
        for (int i = 0; i < inp.size(); i++) {
            int num = Integer.parseInt(inp.get(i).substring(1));
            int turn;
            switch (inp.get(i).charAt(0))
            {
                case('F'):
                    if(facing==0)
                        y+=num;
                    else if(facing==1)
                        x+=num;
                    else if(facing==2)
                        y-=num;
                    else if(facing==3)
                        x-=num;
                    break;
                case('R'):
                    turn = degreetonum(num);
                    facing+=turn;
                    if(facing>3)
                        facing-=4;
                    break;
                case('L'):
                    turn = degreetonum(num);
                    facing-=turn;
                    if(facing<0)
                        facing+=4;
                    break;
                case('N'):
                    y+=num;
                    break;
                case('S'):
                    y-=num;
                    break;
                case('E'):
                    x+=num;
                    break;
                case('W'):
                    x-=num;
                    break;
            }
        }
        System.out.println("Part1: "+(Math.abs(x)+Math.abs(y)));
        int x1=10,y1=1;
        x=0;
        y=0;
        for (int i = 0; i < inp.size(); i++) {
            int num = Integer.parseInt(inp.get(i).substring(1));
            int turn;
            switch (inp.get(i).charAt(0))
            {
                case('F'):
                    for (int j = 0; j < num; j++) {
                        x+=x1;
                        y+=y1;
                    }
                    break;
                case('R'):
                    turn = degreetonum(num);
                    if(turn == 1)
                    {
                        int temp = x1;
                        x1=y1;
                        y1=-1*temp;
                    }
                    else if(turn == 2)
                    {
                        x1*=-1;
                        y1*=-1;
                    }
                    else if(turn == 3)
                    {
                        int temp=x1;
                        x1=-1*y1;
                        y1=temp;
                    }
                    break;
                case('L'):
                    turn = degreetonum(num);
                    if(turn == 1)
                    {
                        int temp=x1;
                        x1=-1*y1;
                        y1=temp;
                    }
                    else if(turn == 2)
                    {
                        x1*=-1;
                        y1*=-1;
                    }
                    else if(turn == 3)
                    {
                        int temp = x1;
                        x1=y1;
                        y1=-1*temp;
                    }
                    break;
                case('N'):
                    y1+=num;
                    break;
                case('S'):
                    y1-=num;
                    break;
                case('E'):
                    x1+=num;
                    break;
                case('W'):
                    x1-=num;
                    break;
            }
        }
        System.out.println("Part2: "+(Math.abs(x)+Math.abs(y)));
    }

    private int degreetonum(int num)
    {
        if(num==0)
            return 0;
        if(num==90)
            return 1;
        if(num==180)
            return 2;
        if(num==270)
            return 3;
        return 0;
    }

}
